src/uct/ib/base/ib_device.c

```diff
@@ -629,23 +639,49 @@ size_t uct_ib_mtu_value(enum ibv_mtu mtu)
     ucs_fatal("Invalid MTU value (%d)", mtu);
 }
 
-uint8_t uct_ib_to_fabric_time(double time)
+uint8_t uct_ib_to_qp_fabric_time(double time)
 {
     double to;
-    long t;
 
     to = log(time / 4.096e-6) / log(2.0);
     if (to < 1) {
         return 1; /* Very small timeout */
-    } else if (to > 30) {
+    } else if ((long)(to + 0.5) >= UCT_IB_FABRIC_TIME_MAX) {
         return 0; /* No timeout */
     } else {
-        t = (long)(to + 0.5);
-        ucs_assert(t >= 1 && t < 31);
-        return t;
+        return (long)(to + 0.5);
     }
 }
 
+uint8_t uct_ib_to_rnr_fabric_time(double time)
+{
+    double time_ms = time * UCS_MSEC_PER_SEC; 
+    double avg_ms;
+    uint8_t t;
+
+    if (time_ms <= uct_ib_qp_rnr_time_ms[0]) {
+        return 1;
+    }
+
+    for (t = 1; t < UCT_IB_FABRIC_TIME_MAX; t++) {
+        if (time_ms <= uct_ib_qp_rnr_time_ms[t]) {
+            avg_ms = (uct_ib_qp_rnr_time_ms[t - 1] +
```

Comment:
what is the idea here, why need to compare with entry [t-1] to decide if return [t] or [t+1]?

Reply:
there are two `if` statements here:
- `if (time_ms <= uct_ib_qp_rnr_time_ms[t]) { ... }` - checks whether we should move and and find more suitable value (false branch) or the `time_ms` value inside the find range (true branch)
- `if (time_ms < avg_ms) { ... }` (where `avg_ms` = the average value between `t-1`th and `t`th values) - tests whether the `time_ms` value inside `(table[t-1], avg_ms)` range or inside `[avg_ms, table[t]]` range

Reply:
suppose time is in [avg_ms, table[t]] range. Why need to return t+1 and not t?

Reply:
Sorry for misunderstanding from me
This is a good question :)
The `uct_ib_qp_rnr_time_ms` table defines table as follows:
- `uct_ib_qp_rnr_time_ms[0]=0.01`
- ...
- `uct_ib_qp_rnr_time_ms[31]=491.52`
- `uct_ib_qp_rnr_time_ms[31]=655.36`

But IBTA specification defines the table as follows:
- `uct_ib_qp_rnr_time_ms[0]=655.36`
- `uct_ib_qp_rnr_time_ms[1]=0.01`
- ...
- `uct_ib_qp_rnr_time_ms[31]=491.52`

We can redefine table as it in IBTA specification.
The `uct_ib_to_rnr_fabric_time` function code will be updated in the following way:
```
uint8_t uct_ib_to_rnr_fabric_time(double time)
{
    double time_ms = time * UCS_MSEC_PER_SEC;
    double avg_ms;
    uint8_t t;

    for (t = 1; t < UCT_IB_FABRIC_TIME_MAX; t++) {
        if (time_ms <= uct_ib_qp_rnr_time_ms[(t + 1) % UCT_IB_FABRIC_TIME_MAX]) {
            avg_ms = (uct_ib_qp_rnr_time_ms[t] +
                      uct_ib_qp_rnr_time_ms[(t + 1) % UCT_IB_FABRIC_TIME_MAX]) * 0.5;

            if (time_ms < avg_ms) {
                /* return previous value */
                return t;
            } else {
                /* return current value */
                return (t + 1) % UCT_IB_FABRIC_TIME_MAX;
            }
        }
     }

    return 0; /* this is a special value that means the maximum value */
 }
```

@yosefe wdyt? should update to reflect the IBTA spec table (it simplifies `uct_ib_to_rnr_fabric_time`)?

Reply:
yeah. let's update it.
I guess i was confused by the comment which says the table in the code is according to IBTA :)